### PR TITLE
use implied bounds from impl header when comparing trait and impl methods

### PR DESCRIPTION
**Pull request by aliemjay**
Fixes #105495.

I pulled it out of #105483.

r? @compiler-errors

### CODE REVIEW

File: src/test/ui/implied-bounds/compare-impl-method-projections.rs [hunk 1]
```diff
@@ -0,0 +1,47 @@
+// Same as ./compare-impl-method.rs but we shouldn't use implied bounds from projections in impl
+// header. See `mod exploit` for why.
```
Comment (aliemjay) [inline src/test/ui/implied-bounds/compare-impl-method-projections.rs:2]: @compiler-errors It turned out there's a good reason why `ocx.assumed_wf_types` doesn't include unnormalized projections :wink: .

Comment (compiler-errors) [reply src/test/ui/implied-bounds/compare-impl-method-projections.rs:2]: Oh, I'm aware of the current issue with `assumed_wf_types`, but I'm assuming that even more crates will fail if we don't add both unnormalized *and* normalized, but then enforce unnormalized types are WF in #105483.

Though I'm not sure if @lcnr is working on the fix for that they mentioned in the comment in the body of `ObligationCtxt::assumed_wf_types`.

Comment (compiler-errors) [reply src/test/ui/implied-bounds/compare-impl-method-projections.rs:2]: I'm assuming #105483 replaces a large unsoundness hole with a smaller unsoundness hole, so :shrug:

**Comment by lcnr** [comment]
Unless there's another reason apart from consistency for adding this, I think this should wait until we're able to add explicit implied bounds in the trait solver. I think we should block this on coinductive trait predicates.

## What does this PR do

When we check that the `where`-bounds of the impl are weaker than the `where`-bounds of the trait, we now add the implied bounds of the impl header to our assumptions, allowing `where`-bounds which rely on these implied bounds.

## Does this agree with our long term plan for implied bounds

Definitely yes. We pretty much want to get to a point where we add `WellFormed` predicates for types in the impl signature to the `predicates_of` and then use these bounds as assumptions while inside of the item.

This means that any place which uses the `param_env` of the impl should also be able to use the implied bounds of the impl.

## Is this currently sound

Sadly not completely. It would allow for a slightly different variant of https://github.com/rust-lang/rust/issues/100051. Instead of now only relying on this unchecked implied bounds when checking that the *signature* has weaker requirements, we also rely on them when checking the *`where`-bounds*.

This causes the following unsound test to compile:

(pretty much https://github.com/rust-lang/rust/issues/100051)
```rust
trait Trait {
    type Type;
}

impl<T> Trait for T {
    type Type = ();
}

trait Extend<'a, 'b> {
    fn extend(s: &'a str) -> &'b str;
}

impl<'a, 'b> Extend<'a, 'b> for <&'b &'a () as Trait>::Type
where
    for<'what, 'ever> &'what &'ever (): Trait,
{
    // instead of getting the implied bound from the self type
    // we get it through an explicit bound on the impl method instead.
    //
    // This explicit bound is accepted because its implied by the
    // impl header.
    fn extend(s: &'a str) -> &'b str
    where
        'a: 'b
    {
        s
    }
}

fn main() {
    let y = <() as Extend<'_, '_>>::extend(&String::from("Hello World"));
    println!("{}", y);
}
```

I don't think this unsoundness will ever get triggered by accident and if someone wants to use it, they can already use the https://github.com/rust-lang/rust/issues/100051. So I don't think that this is a hard blocker to landing this PR if there were practical applications.

## Is it useful

I don't think it's too useful. If bounds on the impl method rely on the implied bounds from the impl header to be correct, then why add these as bounds at all if they're already implied by the impl header.

borrowck also doesn't use implied bounds from the impl header right now, only from its signature, so this PR would allow circumventing that issue. That would be pretty inconsistent though. Allowing borrowck to use implied bounds from impl headers would allow the above unsoundness in a more implicit way so I fear that it might actually be possible to rely on that by accident.

---

I guess let's go with:

@rust-lang/types I believe we should delay adding implied bounds to `compare_predicate_entailment` until we're correctly dealing with implied bounds from unnormalized projections in the impl header.

@rfcbot fcp close

**Comment by lcnr** [comment]
@rfcbot fcp cancel

**Comment by lcnr** [comment]
@rfcbot fcp close

**Pull request closed without merging**